scipy/stats/_qmc_cy.pyx

```diff
@@ -344,18 +413,62 @@ def _cy_van_der_corput(long n, long base, long start_index):
             b2r /= base
             quotient //= base
 
-    return sequence
 
 
-def _cy_van_der_corput_scrambled(long n, long base, long start_index,
-                                 long[:, ::1] permutations):
+cdef _cy_van_der_corput_scrambled(Py_ssize_t n,
```

Comment:
Why do we need to have this function?

Reply:
As mentioned in my previous comment, on the Cython's documentation _The Performance of Python, Cython and C on a Vector_ (link [here](https://notes-on-cython.readthedocs.io/en/latest/std_dev.html)), the fatest execution of Cython code is when you have a python `def` function, which wraps a`cdef` function. So I just repeated this pattern here to achieve the maximum

Reply:
Yes but `_cy_van_der_corput_scrambled` is a `cdef` and `_cy_van_der_corput_scrambled_loops` is also a `cdef`. So I don't get why the need of this wrapping.

Reply:
I wanted to be more like the pattern used in the discrepancy Cython code, but I refactorised it in 81ca9f5e3, and I think it is indeed better. Does it suit you?

Reply:
Yes better thanks. I am still wondering if we could remove the wrapper part as well. The part which is called multiple times needs to be a `cdef`, but the rest can be `cpdef`. Because in the end you are still calling a `def` which calls a `cdef`. This is what cython do internally so IMO we should change this.

Reply:
sorry, actually it is this article which explains which is the fastest way to define function : 
https://notes-on-cython.readthedocs.io/en/latest/fibo_speed.html

Reply:
Yes exactly, in our case `_cy_van_der_corput_scrambled` is only called once. The way I see this is we could have `cpdef _cy_van_der_corput_scrambled` which is calling multiple `cdef _cy_van_der_corput_scrambled_loop`. It's not a big deal, but if you could do both `timeit`, that would be enlightening.

Reply:
1. With `cdef` and a `def` function wrapper : 

|` n=100`  | ` n=500` | ` n=1000` | ` n=10_000` | ` n=100_000` |
|-----------|----------|-----------|-------------|--------------|
|  8.5µs | 49.9µs | 110.8µs |  1.6ms | 19.7ms |

2. With only one `cpdef` function (`_cy_van_der_corput`, without `wrapper`) 

|` n=100`  | ` n=500` | ` n=1000` | ` n=10_000` | ` n=100_000` |
|-----------|----------|-----------|-------------|--------------|
|  9.7µs | 53.8µs | 112.1µs |  1.6ms | 20.4ms |


3. Same one as previously but with `def` (`_cy_van_der_corput`, without `wrapper`, define as a `def`) 
|` n=100`  | ` n=500` | ` n=1000` | ` n=10_000` | ` n=100_000` |
|-----------|----------|-----------|-------------|--------------|
|  8.2µs | 50.0µs | 111.3µs |  1.6ms | 20.7ms |

4. with defining `_cy_wrapper_van_der_corput` with `cpdef` and `cdef` `_cy_van_der_corput`
|` n=100`  | ` n=500` | ` n=1000` | ` n=10_000` | ` n=100_000` |
|-----------|----------|-----------|-------------|--------------|
|  9.9µs | 55.3µs | 114.1µs |  1.6ms | 21.2ms |

I think we are for 1. and 3. in the IC. So up to you on the one you want

Reply:
All of these give very similar results. Then let's go with no wrapper and either a def or cpdef calling the cdef loop. It just simply the stack. Thanks for doing the analysis!

Reply:
Sorry for the delay. Changes are in e8e33208a. Btw, is the comment on the threaded part clear enough for you? (Line 17 to 22 in `_qmc_cy.pyx`)

Reply:
All good now, thanks! Yes I already closed this conversation it's clear for me.